omas/omas_physics.py

```diff
@@ -203,6 +203,131 @@ def core_profiles_zeff(ods, update=True, use_electrons_density=False):
+    elif 'j_bootstrap' in prof1d:
+        # update j_non_inductive with latest bootstrap current
+        if 'j_non_inductive' in prof1d:
+            prof1d['j_non_inductive'] += prof1d['j_bootstrap']
+            if 'j_bootstrap' in j_old:
+                prof1d['j_non_inductive'] -= j_old['j_bootstrap']
+        else:
+            prof1d['j_non_inductive'] = prof1d['j_bootstrap']
+
+    # Total parallel current
+    if j_total is not None:
+        # use the provided current
+        prof1d['j_total'] = j_total
+    else:
+        # update total current with latest currents
+        for j in ['j_ohmic', 'j_non_inductive']:
+            if j in prof1d:
+                if 'j_total' in prof1d:
+                    prof1d['j_total'] += prof1d[j]
+                    if j in j_old:
+                        prof1d['j_total'] -= j_old[j]
+                else:
+                    prof1d['j_total'] = prof1d[j]
+
+    # get some quantities we'll use below
+    eq = ods['equilibrium']['time_slice'][time_index]
+    if 'core_profiles.vacuum_toroidal_field.b0' in ods:
+        B0 = ods['core_profiles']['vacuum_toroidal_field']['b0'][time_index]
+    elif 'equilibrium.vacuum_toroidal_field.b0' in ods:
+        B0 = ods['equilibrium']['vacuum_toroidal_field']['b0'][time_index]
+    rho = prof1d['grid']['rho_tor_norm']
+
+    # Total toroidal current
+    if j_tor is not None:
+        # use the provided current
+        prof1d['j_tor'] = j_tor
+    elif 'j_total' in prof1d:
+        # update toroidal current using transformation
+        JparB_tot = prof1d['j_total']*B0
+        JtoR_tot = transform_current(rho, JparB=JparB_tot,
+                                     equilibrium=eq, includes_bootstrap=True)
+        fsa_invR = numpy.interp(rho, eq['profiles_1d']['rho_tor_norm'], eq['profiles_1d']['gm9'])
+        prof1d['j_tor'] = JtoR_tot/fsa_invR
+
+
+    # Calculate integrated currents
+    rho_eq   = eq['profiles_1d']['rho_tor_norm']
+    vp       = eq['profiles_1d']['dvolume_dpsi']
+    psi      = eq['profiles_1d']['psi']
+    fsa_invR = eq['profiles_1d']['gm9']
+    with omas_environment(ods,
+                          coordsio={'core_profiles.profiles_1d.%d.grid.rho_tor_norm'%time_index:rho_eq}):
+
+        currents = [('j_bootstrap', 'current_bootstrap', True),
+                    ('j_non_inductive', 'current_non_inductive', True),
+                    ('j_tor', 'ip', False)]
+
+        for Jname, Iname, transform in currents:
+            J = prof1d[Jname]
+            if transform:
+                # transform <J.B>/B0 to <Jt/R>
+                J = transform_current(rho_eq, JparB=J*B0,
+                                      equilibrium=eq, includes_bootstrap=True)
+            else:
+                # already <Jt/R>/<1/R>
+                J *= fsa_invR
+
+            ods.set_time_array('core_profiles.global_quantities.%s'%Iname,time_index,
+                               cumtrapz(vp*J,psi)[-1]/(2.*numpy.pi))
```

Comment:
@bclyons12 I may be misunderstanding something in your calculation of the integrated currents, but shouldn't it be like this?

```python
ar=eq['profiles_1d']['darea_dpsi']
...
ods.set_time_array('core_profiles.global_quantities.%s'%Iname,time_index, trapz(ar*J,psi))

```

Reply:
So `J` here is actually `<Jt/R>`. The `psi` integral turns the flux-surface average into a full volume integral, so you want the volume differential. I've derived this myself before, but don't remember the details off the top of my head. It's the same thing that's done in `OMFITgeqdsk.calc_Ip` though.